triton/tail.go

```diff
@@ -0,0 +1,249 @@
+		stream:            params.StreamName,
+		bucket:            params.Bucket,
+		client:            params.Client,
+		records:           make(chan Record),
+		errors:            make(chan error),
+	}
+	return
+}
+
+func (t *TailAt) Next() (record Record, err error) {
+	t.Lock()
+	defer t.Unlock()
+	go t.initStream()
+	select {
+	case record = <-t.records:
+	case err = <-t.errors:
+	}
+	return
+}
+
+// sendArchivedRecords sends all the archived records between tail at and up to a day afterward.
+func (t *TailAt) sendArchivedRecords() (lastMetadata *StreamMetadata, err error) {
+	archiveRepository := NewArchiveRepository(t.s3Service, nil, t.bucket, t.stream, t.client)
+	aTime := t.at.AddDate(0, 0, -1)
+	end := t.at.AddDate(0, 0, 2)
+	var lastArchive *StoreArchive
+	for aTime.Before(end) {
+		if t.closed {
+			break
+		}
+		if !aTime.Before(end) {
+			break
+		}
+		var archives []StoreArchive
+		archives, err = archiveRepository.ArchivesAtDate(aTime)
+		if err != nil {
+			return
+		}
+		for _, archive := range archives {
+			if t.closed {
+				return
+			}
+			lastArchive = &archive
+			if archive.T.Before(t.at) {
+				continue
+			}
+			for {
+				var rec map[string]interface{}
+				rec, err = archive.ReadRecord()
+				if err == io.EOF {
+					err = nil
+					break
+				} else if err != nil {
+					return
+				}
+				t.records <- rec
+			}
+		}
+		aTime = aTime.AddDate(0, 0, 1)
+	}
+	if lastArchive != nil {
+		lastMetadata, err = lastArchive.GetStreamMetadata()
+	}
+	return
+}
+
+// initStream is internal method that starts sending archived records followed
+// by reading from Kinesis shards in parallel
+func (t *TailAt) initStream() {
+	if t.streamInited {
+		return
+	}
+	t.streamInited = true
+
+	lastStreamMetadata, err := t.sendArchivedRecords()
+	if err != nil {
+		t.errors <- err
+		return
+	}
+	err = t.sendKinesisRecords(lastStreamMetadata)
+	if err != nil {
+		t.errors <- err
+	}
+}
+
+func (t *TailAt) sendKinesisRecords(previousMetadata *StreamMetadata) (err error) {
+	shards, err := t.listShards()
+	if err != nil {
+		return
+	}
+
+	// send all of the records in `startingKey`
+	// load metadata for starting key
+	// then send kinesis records
+	// load the sequenceNumbers for the last key
+	for _, shard := range shards {
+		var lastSequenceNumber SequenceNumber
+		if previousMetadata != nil && previousMetadata.Shards[shard] != nil {
+			lastSequenceNumber = previousMetadata.Shards[shard].MaxSequenceNumber
+		}
+		go t.sendKinesisRecordsForShard(shard, lastSequenceNumber)
+	}
+	return
+}
+
+// sendKinesisRecordsForShard starts sending records to TailAt.records for the shard optionally starting at starting startingSequenceNumber
+//
+// If a startingSequenceNumber is specified the reader tries to start reading
+// records at startingSequenceNumber, otherwise it tries to find a starting
+// sequence number at TRIM_HORIZON to begin reading.
+func (t *TailAt) sendKinesisRecordsForShard(shard ShardID, startingSequenceNumber SequenceNumber) {
```

Comment:
Can you replace this with a StreamReader... maybe enhancing the interfaces there to do what you want? (start at a specific sequence number)

The existing `NewStreamReader(svc KinesisService, streamName string, c Checkpointer)` might not work because Checkpointer doesn't do exactly what you want? But i think something could be added to make this work.